### ISSUE TITLE
Blocks / transactions propagation improvements

### ISSUE THREAD

**Issue by gumb0** [ethereum/aleth#5277]
From the recent discussion in AllCoreDevs channel https://gitter.im/ethereum/AllCoreDevs?at=5b9d1bf0b4990c30eebcba0f it looks like the good strategy for selecting the peers for propagation would be
1. Block propagation - send block to square root of number of peers who do not have that block yet, but limit with minimum 4 peers
> Propagating to all peers would lead to O(nm) messages being sent, where n is the number of nodes and m the number of peers per node - this devolves to O(n^2) in a fully connected network.
If we propagate to sqrt(m) peers, then in a fully connected network we send O(n * sqrt(n)) messages, which grows much slower.

2. Transactions - propagate to all peers

> the reason transactions are propagated to all peers is to avoid gaps
if we start propagating transactions to sqrt peers, we'll fill up every txpool with nonce gaps, causing transactions to be dumped

What we do:
1. Propagate blocks to 25% of all peers (but only to those who do not have that block yet)
https://github.com/ethereum/aleth/blob/1114b251b8e09d3631557debee3ca1d40f002fa6/libethereum/EthereumHost.cpp#L554-L558
2. Propagate transaction to all peers who don't know it yet
https://github.com/ethereum/aleth/blob/1114b251b8e09d3631557debee3ca1d40f002fa6/libethereum/EthereumHost.cpp#L464

So 1 can be improved; 2 is fine, but using `randomSelection` there looks very weird, we should select peers without it.

**Comment by gumb0** [comment]
Further thoughts on block propagation, can be done as separate tasks, might be non-trivial changes:

3. In case we're mining it makes sense to propagate the blocked mines by us to 100% of peers.

4. There's a consensus now that it's more efficient for the network as a whole to propagate the blocks as soon as the PoW check is complete. go-ethereum already does that, Parity plans to do that, we still propagate only after the blocks are fully validated and imported into the chain.

**Comment by halfalicious** [comment]
>     1. There's a consensus now that it's more efficient for the network as a whole to propagate the blocks as soon as the PoW check is complete. go-ethereum already does that, Parity plans to do that, we still propagate only after the blocks are fully validated and imported into the chain.

From the perspective of reduced latency between a new block being mined and it reaching all peers that makes sense. Some questions related to this approach:
* A downside to propagating blocks before they're fully validated is the possibility of propagating blocks with invalid transactions which will result in unnecessary network traffic. Presumably this isn't an issue because we expect such blocks to be rare - there's no incentive for a miner to try and propagate such an invalid block since it will eventually be detected (when each node imports the block), which means it won't be included in the chain and therefore they won't get the block reward and will have wasted the energy used to compute the PoW? 
* Do nodes do anything beyond dropping invalid blocks when they're detected during chain import e.g. do they warn peers about the bad block?

**Comment by gumb0** [comment]
> A downside to propagating blocks before they're fully validated is the possibility of propagating blocks with invalid transactions which will result in unnecessary network traffic. Presumably this isn't an issue because we expect such blocks to be rare - there's no incentive for a miner to try and propagate such an invalid block since it will eventually be detected (when each node imports the block), which means it won't be included in the chain and therefore they won't get the block reward and will have wasted the energy used to compute the PoW?

@halfalicious Right, you can try to waste all validating nodes' resources by flooding the network with invalid blocks, but it's expensive to do so (need to compute PoW for each)

> Do nodes do anything beyond dropping invalid blocks when they're detected during chain import e.g. do they warn peers about the bad block?

There's no warning of other peers. The node that sent the bad block may be recorded internally by the node as having bad reputation.

Geth also saves the bad blocks for debugging purposes, we just discard them (and lower peer's rating I think)

**Comment by gumb0** [comment]
Created a separate issue for point 4 https://github.com/ethereum/aleth/issues/5371

**Comment by realisation** [comment]
About how long does it take for a mined block to be propagated through the whole ethereum network? How long for lite nodes and how long for full nodes?
